File: ios-native/NeedsVsWants/Data/EntryRepository.swift
import Foundation
import SwiftData

enum RepositoryError: Error {
    case saveFailed
    case deleteFailed
}

/// All data access goes through here.
///
/// Improvement over the old `ios/` port's god `EntryStore`: this is a thin,
/// `Result`-returning CRUD layer — no `try?` swallowing errors, no stats
/// aggregation mixed in (that lives in `StatsEngine`). Views read via `@Query`;
/// they do not call the repository for reads.
@MainActor
final class EntryRepository {
    private let context: ModelContext

    init(context: ModelContext) {
        self.context = context
    }

    // MARK: - Writes

    /// Insert an entry stamped at now. Sealing IS saving (D5).
    @discardableResult
    func insert(item: String, costCents: Int64, type: EntryType) -> Result<Entry, RepositoryError> {
        insertAt(date: Date(), item: item, costCents: costCents, type: type)
    }

    /// Insert an entry at an explicit date (used by tests and back-dating).
    @discardableResult
    func insertAt(date: Date, item: String, costCents: Int64, type: EntryType) -> Result<Entry, RepositoryError> {
        let entry = Entry(dateUtc: date, item: item, costCents: costCents, type: type)
        context.insert(entry)
        do {
            try context.save()
            return .success(entry)
        } catch {
            return .failure(.saveFailed)
        }
    }

    @discardableResult
    func delete(_ entry: Entry) -> Result<Void, RepositoryError> {
        context.delete(entry)
        do {
            try context.save()
            return .success(())
        } catch {
            return .failure(.deleteFailed)
        }
    }

    @discardableResult
    func deleteAll() -> Result<Void, RepositoryError> {
        let all = (try? context.fetch(FetchDescriptor<Entry>())) ?? []
        for e in all { context.delete(e) }
        do {
            try context.save()
            return .success(())
        } catch {
            return .failure(.saveFailed)
        }
    }

    // MARK: - Retention (D3): silent auto-purge, no prompt.

    func purgeOlderThan(days: Int) {
        guard let cutoff = Calendar.current.date(byAdding: .day, value: -days, to: Date()) else { return }
        let predicate = #Predicate<Entry> { $0.dateUtc < cutoff }
        if let stale = try? context.fetch(FetchDescriptor<Entry>(predicate: predicate)) {
            for e in stale { context.delete(e) }
            try? context.save()
        }
    }

    // MARK: - Sheet cap (D4 / Android parity: counts TOTAL entries)

    var sheetCount: Int { (try? context.fetch(FetchDescriptor<Entry>()))?.count ?? 0 }
    var isSheetFull: Bool { sheetCount >= 20 }

    // MARK: - Stats

    func stats(for period: Period, engine: StatsEngine) -> SummaryStats {
        let entries = (try? context.fetch(FetchDescriptor<Entry>())) ?? []
        return engine.stats(for: period, entries: entries)
    }
}

File: expo/src/data/EntryRepository.ts
/**
 * Entry repository — data access layer.
 *
 * Faithful port of EntryRepository.swift. All writes return Result-style
 * booleans (no silent swallowing). Reads return plain Entry objects.
 *
 * Key behaviors preserved:
 *  - insert() stamps dateUtc = Date.now() (local TZ moment)
 *  - purgeOlderThan(35) silently deletes entries older than 35 days
 *  - sheetCount counts TOTAL entries (not per-day); isSheetFull at >= 20
 *  - stats() delegates to StatsEngine
 */
import { type SQLiteDatabase } from "expo-sqlite";
import { EMPTY_STATS, type Entry, type Period, type SummaryStats } from "./schema";
import { StatsEngine } from "./StatsEngine";

const SHEET_CAP = 20;
const RETENTION_DAYS = 35;
const MS_PER_DAY = 24 * 60 * 60 * 1000;

interface EntryRow {
  id: string;
  dateUtc: number;
  item: string;
  costCents: number;
  typeRaw: "NEED" | "WANT";
}

function rowToEntry(row: EntryRow): Entry {
  return { ...row };
}

export class EntryRepository {
  constructor(private readonly db: SQLiteDatabase) {}

  async insert(item: string, costCents: number, type: "NEED" | "WANT"): Promise<boolean> {
    try {
      const entry: Entry = {
        id: crypto.randomUUID(),
        dateUtc: Date.now(),
        item,
        costCents,
        typeRaw: type,
      };
      await this.db.runAsync(
        "INSERT INTO entries (id, dateUtc, item, costCents, typeRaw) VALUES (?, ?, ?, ?, ?)",
        [entry.id, entry.dateUtc, entry.item, entry.costCents, entry.typeRaw],
      );
      return true;
    } catch {
      return false;
    }
  }

  /** Explicit-date insert (for back-dating / tests). */
  async insertAt(dateUtc: number, item: string, costCents: number, type: "NEED" | "WANT"): Promise<boolean> {
    try {
      const entry: Entry = {
        id: crypto.randomUUID(),
        dateUtc,
        item,
        costCents,
        typeRaw: type,
      };
      await this.db.runAsync(
        "INSERT INTO entries (id, dateUtc, item, costCents, typeRaw) VALUES (?, ?, ?, ?, ?)",
        [entry.id, entry.dateUtc, entry.item, entry.costCents, entry.typeRaw],
      );
      return true;
    } catch {
      return false;
    }
  }

  async delete(id: string): Promise<boolean> {
    try {
      await this.db.runAsync("DELETE FROM entries WHERE id = ?", [id]);
      return true;
    } catch {
      return false;
    }
  }

  async deleteAll(): Promise<boolean> {
    try {
      await this.db.runAsync("DELETE FROM entries");
      return true;
    } catch {
      return false;
    }
  }

  /** Silently deletes entries older than `days`. Called once per launch. */
  async purgeOlderThan(days: number = RETENTION_DAYS): Promise<void> {
    const cutoff = Date.now() - days * MS_PER_DAY;
    await this.db.runAsync("DELETE FROM entries WHERE dateUtc < ?", [cutoff]);
  }

  /** Total entry count across the diary (not per-day). */
  async sheetCount(): Promise<number> {
    const row = await this.db.getFirstAsync<{ count: number }>("SELECT COUNT(*) as count FROM entries");
    return row?.count ?? 0;
  }

  async isSheetFull(): Promise<boolean> {
    return (await this.sheetCount()) >= SHEET_CAP;
  }

  /** All entries, newest first. */
  async all(): Promise<Entry[]> {
    const rows = await this.db.getAllAsync<EntryRow>(
      "SELECT * FROM entries ORDER BY dateUtc DESC",
    );
    return rows.map(rowToEntry);
  }

  async stats(period: Period, engine: StatsEngine): Promise<SummaryStats> {
    const entries = await this.all();
    return engine.stats(period, entries);
  }

  get sheetCap(): number {
    return SHEET_CAP;
  }
}

export { SHEET_CAP, RETENTION_DAYS, EMPTY_STATS };
